### DISCUSSION TITLE
A11y Warnings

### DISCUSSION THREAD

**Discussion by lorenzolewis** [withastro/roadmap#450] [Proposal]
# Summary

Add a11y (accessibility) warnings to the Astro compiler similar to that of Svelte (https://svelte.dev/docs#accessibility-warnings)

# Background & Motivation

Let's face it: a11y can be hard. I really enjoyed using the warnings built into the Svelte compiler to guide me towards common a11y issues. This has personally helped me a lot in learning about the different features built into HTML that can help create a more accessible web for everyone.

# Goals

- Add in compiler warnings for common scenarios around a11y issues
- Add a way to silence those warnings on a per-case basis (if you're aware and want to ignore)
- Initially start with just Astro components. I'm not sure if this would be possible to add outside, but that could be a thought exercise for later.

**Comment by natemoo-re** [comment]
Thanks for the proposal @lorenzolewis! This is something I've wanted to add to the Astro compiler for a long time. We have all the infrastructure in place to surface warnings/errors, we just need to wire up the actual a11y logic.

I'd be really excited to work on this!

**Comment by matthewp** [comment]
I think this is something we could do. Definitely do agree that limiting to `.astro` is a good choice to keep the scope down. What I would be interested in is

1. Can you provide a list of rules we should have?
2. What is the mechanism you are thinking for how we allow disabling a rule? Using comments?
3. When/where would these checks run? Via `astro check`?

**Reply by Princesseuh** [reply]
For 3., I would assume that this would work the same way the current compiler warnings do, so in the editor and in `astro check`

**Reply by lorenzolewis** [reply]
> 1. Can you provide a list of rules we should have?

I had a quick re-read through Svelte's docs and I think all of them there are pretty good as a first pass (https://svelte.dev/docs#accessibility-warnings-a11y-autofocus). I'm assuming they created that iteratively since a lot of A11Y rules aren't necessarily straightforward and easy to distill down into code-able logic

> 2. What is the mechanism you are thinking for how we allow disabling a rule? Using comments?

I'd suggest comments, but I've had a bit of issues with comments within complex Astro components (where Command-slash will put in a JS comment when it should be HTML or vice-versa). If it is comments then I don't think you want those to be passed to the output HTML like all HTML comments currently are (from my recollection). So that will be a different DX

> 3. When/where would these checks run? Via astro check?

Echoing @Princesseuh's comment here, compiler warnings and also `astro check`

**Comment by florian-lefebvre** [comment]
You can now get a11y warning through the dev toolbar: https://docs.astro.build/en/guides/dev-toolbar/#audit
